node/pkg/watchers/evm/watcher.go

```diff
@@ -279,25 +279,57 @@ func (w *Watcher) Run(ctx context.Context) error {
-			p2p.DefaultRegistry.AddErrorCount(w.chainID, 1)
-			return fmt.Errorf("dialing eth client failed: %w", err)
-		}
-		finalizer, err := finalizers.NewOptimismFinalizer(timeout, logger, w.l1Finalizer, w.rootChainRpc, w.rootChainContract)
-		if err != nil {
-			p2p.DefaultRegistry.AddErrorCount(w.chainID, 1)
-			return fmt.Errorf("creating optimism finalizer failed: %w", err)
-		}
-		w.ethConn, err = connectors.NewBlockPollConnector(ctx, baseConnector, finalizer, 250*time.Millisecond, false, false)
-		if err != nil {
-			ethConnectionErrors.WithLabelValues(w.networkName, "dial_error").Inc()
-			p2p.DefaultRegistry.AddErrorCount(w.chainID, 1)
-			return fmt.Errorf("creating block poll connector failed: %w", err)
+		if w.rootChainRpc != "" && w.rootChainContract != "" {
+			// We are in pre-Bedrock mode
+			if w.l1Finalizer == nil {
+				return fmt.Errorf("unable to create optimism watcher because the l1 finalizer is not set")
+			}
+			baseConnector, err := connectors.NewEthereumConnector(timeout, w.networkName, w.url, w.contract, logger)
+			if err != nil {
+				ethConnectionErrors.WithLabelValues(w.networkName, "dial_error").Inc()
+				p2p.DefaultRegistry.AddErrorCount(w.chainID, 1)
+				return fmt.Errorf("dialing eth client failed: %w", err)
+			}
+			finalizer, err := finalizers.NewOptimismFinalizer(timeout, logger, w.l1Finalizer, w.rootChainRpc, w.rootChainContract)
+			if err != nil {
+				p2p.DefaultRegistry.AddErrorCount(w.chainID, 1)
+				return fmt.Errorf("creating optimism finalizer failed: %w", err)
+			}
+			w.ethConn, err = connectors.NewBlockPollConnector(ctx, baseConnector, finalizer, 250*time.Millisecond, false, false)
+			if err != nil {
+				ethConnectionErrors.WithLabelValues(w.networkName, "dial_error").Inc()
+				p2p.DefaultRegistry.AddErrorCount(w.chainID, 1)
+				return fmt.Errorf("creating block poll connector failed: %w", err)
+			}
+		} else {
+			// Wea are in Bedrock mode
+			useFinalizedBlocks = !w.unsafeDevMode
+			safeBlocksSupported := false
+			if useFinalizedBlocks {
+				safeBlocksSupported = true
+				logger.Info("using finalized blocks, will publish safe blocks")
+			} else {
+				logger.Info("using finalized blocks")
+			}
+			baseConnector, err := connectors.NewEthereumConnector(timeout, w.networkName, w.url, w.contract, logger)
+			if err != nil {
+				ethConnectionErrors.WithLabelValues(w.networkName, "dial_error").Inc()
+				p2p.DefaultRegistry.AddErrorCount(w.chainID, 1)
+				return fmt.Errorf("dialing eth client failed: %w", err)
+			}
+			pollConnector, err := connectors.NewBlockPollConnector(ctx, baseConnector, finalizers.NewDefaultFinalizer(), 250*time.Millisecond, useFinalizedBlocks, safeBlocksSupported)
+			if err != nil {
+				ethConnectionErrors.WithLabelValues(w.networkName, "dial_error").Inc()
+				p2p.DefaultRegistry.AddErrorCount(w.chainID, 1)
+				return fmt.Errorf("creating block poll connector failed: %w", err)
+			}
+			// I know this says Arbitrum.  That's just what the type is called.
```

Comment:
Please add a comment saying we need this for `TimeOfBlockByHash`.

Reply:
Fixed.